src/Form/Extension/AutomaticRedirectTypeExtension.php

```diff
@@ -44,19 +55,50 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                     'class' => 'js-add-automatic-redirection-checkbox',
                 ],
             ]);
+
+            $this->oldSlugs[self::getObjectHash($data)] = $data->getSlug();
         });
 
         $builder->addEventListener(FormEvents::POST_SUBMIT, function (FormEvent $event): void {
             $form = $event->getForm();
+
+            /** @var SlugAwareInterface $data */
             $data = $event->getData();
 
             if (!$form->has(self::FIELD_NAME)) {
                 return;
             }
 
-            if ($form->get(self::FIELD_NAME)->getData() === true) {
-                $this->automaticRedirectCreationDecider->askAutomaticRedirectCreation($data);
+            $hash = self::getObjectHash($data);
+            if (!isset($this->oldSlugs[$hash])) {
+                return;
+            }
+
+            $oldSlug = $this->oldSlugs[$hash];
+            $newSlug = $data->getSlug();
+
+            if (null === $oldSlug || null === $newSlug || $oldSlug === $newSlug) {
+                return;
+            }
+
+            // the automatic redirect creation is not requested by the user
+            if ($form->get(self::FIELD_NAME)->getData() === false) {
+                return;
+            }
```

Comment:
I would first check if the automatic redirect creation was asked. It might be better for perfs/logic